redis/src/main/kotlin/rys/ajaxpetproject/redis/repository/CacheMessageRepository.kt

```diff
@@ -0,0 +1,87 @@
+package rys.ajaxpetproject.redis.repository
+
+import org.slf4j.LoggerFactory
+import org.springframework.beans.factory.annotation.Qualifier
+import org.springframework.context.annotation.Primary
+import org.springframework.data.redis.core.ReactiveRedisOperations
+import org.springframework.data.redis.core.ScanOptions
+import org.springframework.stereotype.Repository
+import reactor.core.publisher.Flux
+import reactor.core.publisher.Mono
+import reactor.kotlin.core.publisher.switchIfEmpty
+import reactor.kotlin.core.publisher.toMono
+import rys.ajaxpetproject.internalapi.RedisPrefixes.MESSAGE_CACHE_KEY_PREFIX
+import rys.ajaxpetproject.model.MongoMessage
+import rys.ajaxpetproject.repository.MessageRepository
+
+@Repository
+@Primary
+class CacheMessageRepository(
+    @Qualifier("messageRepositoryImpl") private val actualRepository: MessageRepository,
+    private val redisOperations: ReactiveRedisOperations<String, MongoMessage>
+) : MessageRepository {
+
+    override fun findMessageById(id: String): Mono<MongoMessage> {
+        return redisOperations.opsForValue().get("$MESSAGE_CACHE_KEY_PREFIX$id")
+            .switchIfEmpty {
+                actualRepository.findMessageById(id)
+                    .flatMap { savedMessage ->
+                        redisOperations.opsForValue().set("$MESSAGE_CACHE_KEY_PREFIX$id", savedMessage)
+                            .thenReturn(savedMessage)
+                    }.doOnSuccess { logger.info("Message with id {} was saved in cache", id) }
+            }
+    }
+
+    override fun save(message: MongoMessage): Mono<MongoMessage> {
+        return actualRepository.save(message)
+            .flatMap { savedMessage ->
+                redisOperations.opsForValue().set("$MESSAGE_CACHE_KEY_PREFIX${savedMessage.id!!}", savedMessage)
+                    .doOnSuccess { logger.info("Message with id {} was saved in cache", message.id) }
+                    .thenReturn(savedMessage)
+            }
+    }
+
+    override fun deleteAll(): Mono<Unit> {
+        return actualRepository.deleteAll()
+            .thenMany(
+                redisOperations.scan(ScanOptions.scanOptions().match("$MESSAGE_CACHE_KEY_PREFIX*").build())
+            )
+            .flatMap {
+                redisOperations.opsForValue().delete(it)
```

Comment:
Am I right that Redis is pretty fast so we can perform 1 by 1 deletion? 
(I mean in mongo it was bad practice - creating N delete requests)

Reply:
redis is fast, but think how many network time you'll use